ESBMC Solidity support

kunjsong01 opened the pull request:
Added a new Solidity frontend to enable SMT-based verification of Solidity smart contracts.

This PR replaced ESBMC-Solidity #493 . 

1. JSON lib has been removed by BFG in all commits. 
2. Rebased feature branch and removed the duplicate commits. 
3. Similar to PR ESBMC-Solidity #493, this PR also preserved the original commit history. 
4. Fixed the code based on Rafael's comments:
 * Added json lib in external dependencies using FetchContent
 * Added the option "ENABLE_SOLIDITY_FRONTEND" 


![image](https://user-images.githubusercontent.com/32592800/134078409-1d7079f0-4c23-44ea-bcaa-7fffc0cf2198.png)

kunjsong01 committed:
Adding solidity ast frontend placeholder

Signed-off-by: Kunjian Song <kunjian90@gmail.com>

kunjsong01 committed:
Fixed linking and extension error

kunjsong01 committed:
Lifting barrier in parse() function

kunjsong01 committed:
Added code placeholders for Solidity's ast_convert, parse_treet, declarationt, symbolt, exprt and irept

kunjsong01 committed:
Fixed issues in expr, irep, symbol and added location

kunjsong01 committed:
reverted 9a6501bf0 and 95b31232

kunjsong01 committed:
added -A option for Solidity plain text AST

kunjsong01 committed:
store AST text and json contents in stream objects

kunjsong01 committed:
improved encapsulation by adding libUtil in solidity frontend

kunjsong01 committed:
ported SIF utilities, AST nodes and visitors

kunjsong01 committed:
Fixed embedded nlohmann::nlohmann CMake files

kunjsong01 committed:
embedded SIF into ESBMC

kunjsong01 committed:
added debug prints in libSif/*

kunjsong01 committed:
added context converter placeholder

kunjsong01 committed:
added internal intrinsics and started to convert and add symbols to the context

kunjsong01 committed:
__ESMC_assume: flattened the hierarchical data of clang declarations

kunjsong01 committed:
Added ESBMC_intrinsics object in json to hold all ESBMC or TACAS definitions

kunjsong01 committed:
separate intrincis and ast nodes

kunjsong01 committed:
Added solidity types converter functions and started to process DeclFunction

kunjsong01 committed:
added solidity declaration tracker

kunjsong01 committed:
added json tracker configure file

kunjsong01 committed:
Changed design of decl tracker: use homogeneous tracker instead of a unified namespace

kunjsong01 committed:
added get_type for Builtin type

kunjsong01 committed:
finished q_type parsing in get_type function

kunjsong01 committed:
completed checks for isVariadic and isInlined

kunjsong01 committed:
added get_location_from_decl and get_presumed_location in solidity_convert

kunjsong01 committed:
added set_location and finished locationt setup for DeclFunction

kunjsong01 committed:
completed get_decl_name

kunjsong01 committed:
completed first symbol instantiation

kunjsong01 committed:
added moduleName and the corresponding parsing functions

kunjsong01 committed:
using jsonTrackerRef to improve readibility

kunjsong01 committed:
Successfully added first symbol

kunjsong01 committed:
started to process function params

kunjsong01 committed:
Completed tracker's populate_params

kunjsong01 committed:
improved funcParam struct

kunjsong01 committed:
Completed function param processing

kunjsong01 committed:
Changed -A to -M to hold the dummy main for Solidity verification

kunjsong01 committed:
added member clang_c_module in solidity_ast_language

kunjsong01 committed:
Converted ESBMC intrinsics and dummy main to symbols

kunjsong01 committed:
Extracted ast nodes from json and started to convert

kunjsong01 committed:
clean up

kunjsong01 committed:
restored clang_c_converter

kunjsong01 committed:
Var _x symbol Part I
  - get_type: added get_type, get_sub_type and get_buildin_type

kunjsong01 committed:
Var _x symbol Part II:
  - Completed get_decl_name
  - TODO: Completed get_location

kunjsong01 committed:
Var _x symbol Part III
  - Completed get_decl_location
  - Completed get_default_symbol and added the symbol to the context
  - Completed adjust symbol

kunjsong01 committed:
Function symbol Part I:

  - refactored decl tracker and added inheritance
  - completed get_type
  - completed location
  - completed decl_name
  - added symbol to the context

kunjsong01 committed:
Function symbol Part II:

  - Function args: zero param done
  - BinaryOperator: BO_Assign done (TODO)

kunjsong01 committed:
Function symbol Part II - continued 1

kunjsong01 committed:
Function symbol Part II - continued 2

kunjsong01 committed:
Function symbol Part II - continued 3

  - DeclRefExprClass

kunjsong01 committed:
Function symbol Part II - continued 4

  - ImplicitCastExprClass
  - IntegerLiteralClass

kunjsong01 committed:
Function symbol Part II - continued 4:

  - completed BinaryOperatorExpr symbols for:
    * _x=100;
    * _y=240

kunjsong01 committed:
Function symbol Part III

  - completed BinaryOperatorExpr: BO_Add and BO_Assign for:
    sum = _x + _y;
  - NOTE: removed unnecessary implicit cast to simplfy the process

kunjsong01 committed:
Function symbol Part IV

kunjsong01 committed:
Function symbol Part IV - continued 1:

  - Completed CalleExpr up to args

kunjsong01 committed:
Fix LValueToRValue cast

kunjsong01 committed:
Function symbol Part IV - continued 2:

  - Set up the skeleton for BinaryOpExpr "sum > 100"

kunjsong01 committed:
Function symbol Part IV - continued 3:

  - Completed all symbol conversion

kunjsong01 committed:
Rebased branch to ESBMC latest master

kunjsong01 committed:
Fixed solidity frontend

  - fixed solidity_ast_language module
  - fixed -M argument for dummy main
  - restored Solidity frontend to check overflow
  - added the new msg handler to solidity_converter module

kunjsong01 committed:
Code clean-up Part I:

  - renamed solidity-ast-* to solidity-frontend

kunjsong01 committed:
Code clean-up Part II:

  - removed unused SIF libraries

kunjsong01 committed:
Separate pattern-based check from semantic analysis

kunjsong01 committed:
Code refactoring: overflow - Part I

  - state variable declaration
  - re-wrote conversions based on Solidity grammar rules
    to :
    >> improve the readability of the code
    >> facilitate the maintenance and extension of the code

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch1

  - function definition and body
  - re-wrote conversions based on Solidity grammar rules
    >> added rule parameter-list
    >> added rule function-definition
    >> added rule expression
    >> improved readability, maintenance and extension of the code

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch2

  - make ready for rule expression conversions

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch3

kunjsong01 committed:
Code refactoring: overflow - Part II - Batch4

  - Completing the refactoring!

  Experiment of improvements:
  - Removed all ImplicitCastExpr of the type <IntegralCast>
  - Removed all ImplicitCastExpr of the type <LValueToRValue>
  - CallExpr uses hard-coded "bool" instead of the oiginal "signed_int"
    This is set based on Solidity AST node.

  All experiment passed!

kunjsong01 committed:
Added BO_Sub and BO_LT to verify underflow

kunjsong01 committed:
Stop execution after detecting Tx.Origin vulnerability

kunjsong01 committed:
Fixed the code to use typeDescription instead of typeName

kunjsong01 committed:
Fixed the consistency problem of typeName vs. typeDescriptions

kunjsong01 committed:
Added support for nondet function - Part I

kunjsong01 committed:
Added support for nondet function - Part II

kunjsong01 committed:
Fixed function side effect

kunjsong01 committed:
Fixed function return

kunjsong01 committed:
Added support for nondet function - Part III

kunjsong01 committed:
Added support for nondet function - Part IV

kunjsong01 committed:
Added support for nondet function - Part V

kunjsong01 committed:
Added support for init value of local var declarations

kunjsong01 committed:
Added support for assume function and binary operator != BO_NE

kunjsong01 committed:
Added support for modulo operator %

kunjsong01 committed:
Added support for __VERIFIER_assume

kunjsong01 committed:
Added support for UO_PreDec

kunjsong01 committed:
Added support for ForStatement

kunjsong01 committed:
Fixed ForStmt

kunjsong01 committed:
Added support for array verfication - Part I

  - Fixed duplicate current function name
  - Added support to convert function parameters

kunjsong01 committed:
Added support for array verfication - Part II

kunjsong01 committed:
Added support for array verfication - Part III

  - added code block to handle ArrayTypeName

kunjsong01 committed:
Added support for array verfication - Part IV

  - Added IndexAccess (ArraySubscriptExprClass)
    and ArrayToPointer decay
  - Completed "else" block in if-then-else

kunjsong01 committed:
Code improvements:

  - Removed debug lines
  - Code clean-up
  - Added json lib in external dependencies
  - Added "ENABLE_SOLIDITY_FRONTEND"

lucasccordeiro commented on README.md:
```diff
@@ -7,6 +7,7 @@
 [![codecov](https://codecov.io/gh/esbmc/esbmc/branch/master/graph/badge.svg)](https://codecov.io/gh/esbmc/esbmc)
 
 
+NOTE: This is the dev-solidity-support branch to add Solidity support in ESBMC. This branch was created on 09/03/2021 forked from commit 6c6ae57430859c3 of ESBMC mainline.
```
Is this NOTE still needed?

kunjsong01 replied:
Done. 
Removed this note.

lucasccordeiro commented on src/ansi-c/ansi_c_language.cpp:
```diff
@@ -161,15 +161,15 @@ bool ansi_c_languaget::parse(
   ansi_c_scanner_init();
 
-  bool result = ansi_c_parser.parse();
+  bool result = ansi_c_parser.parse(); // uses yyansi_cparse()
```
Do we need this commented code?

kunjsong01 replied:
Done. 
Removed this comment.

lucasccordeiro commented on src/ansi-c/ansi_c_language.cpp:
```diff
@@ -161,15 +161,15 @@ bool ansi_c_languaget::parse(
     ansi_c_parser.in = &i_preprocessed;
     ansi_c_scanner_init();
-    result = ansi_c_parser.parse();
+    result = ansi_c_parser.parse(); // uses yyansi_cparse()
```
Do we need this commented code?

kunjsong01 replied:
Done.
Removed this comment.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert_literals.cpp:
```diff
@@ -51,7 +51,7 @@ bool clang_c_convertert::convert_integer_literal(
     return true;
 
-  assert(type.is_unsignedbv() || type.is_signedbv());
+  assert(type.is_unsignedbv() || type.is_signedbv()); // for "_x=100", false || true. Same as "assert(sum>100)"
```
Improve this comment.

kunjsong01 replied:
Done. 
Removed this comment. This annotation is redundant as the code is self-explanatory.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert_literals.cpp:
```diff
@@ -63,10 +63,10 @@ bool clang_c_convertert::convert_integer_literal(
       type);
   }
-  else
+  else // "_x=100" uses this. "100" is considered as signed by default. Same as "assert(sum>100)"
```
Improve this comment.

kunjsong01 replied:
Done. 
Removed this comment. This annotation is redundant as the code is self-explanatory.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert_literals.cpp:
```diff
@@ -63,10 +63,10 @@ bool clang_c_convertert::convert_integer_literal(
   {
     the_val = constant_exprt(
-      integer2binary(val.getSExtValue(), bv_width(type)),
+      integer2binary(val.getSExtValue(), bv_width(type)), // val.getSExtValue()=100, bv_width(type)=32
```
Improve this comment.

kunjsong01 replied:
Done. 
Removed this comment. This annotation is redundant as the code is self-explanatory.

lucasccordeiro commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -222,6 +222,15 @@ bool clang_c_languaget::parse(const std::string &path, const messaget &msg)
+  clang_c_convertert converter(context, ASTs, msg);
+  if(converter.convert())
+    return true;
+
```
Remove white space.

kunjsong01 replied:
Done.

lucasccordeiro commented on src/langapi/language_ui.cpp:
```diff
@@ -58,11 +61,20 @@ bool language_uit::parse(const std::string &filename)
   language_filet &lf = result.first->second;
   lf.filename = filename;
-  lf.language = mode_table[mode].new_language(msg);
+  lf.language = mode_table[mode].new_language(msg); // instantiate clang_c_languaget object or other languages' counterpart
```
Improve this comment.

kunjsong01 replied:
Done.
Removed this comment as the code is self-explanatory. This annotation was added when reading the code to follow the verification pipeline, which should be removed in the final patch.

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+{
+  // add more functions here to perform more pattern-based checks
+  //print_json_element(ast_nodes);
+  printf("Checking function %s ...\n", target_func.c_str());
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+      if ((*itr)["kind"].get<std::string>() == "function" &&
+          (*itr)["nodeType"].get<std::string>() == "FunctionDefinition" &&
+          (*itr)["name"].get<std::string>() == target_func)
+      {
```
Remove {}

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+  nlohmann::json::const_iterator itr = body_stmt.begin();
+  for (; itr != body_stmt.end(); ++itr, ++index)
+  {
+    printf("@@ checking body stmt %u\n", index);
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
ACK. I'll replace them with msg.debug(fmt::format(.)).

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+      if ((*itr)["nodeType"].get<std::string>() == "ExpressionStatement")
+      {
+        const nlohmann::json &expr = (*itr)["expression"];
+        //print_json_element(expr);
```
Remove.

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+        {
+          if (expr["kind"] == "functionCall")
+          {
+            // require(.) authorization
```
why?

kunjsong01 replied:
Fixed.
Aplogies for this confusing annotation. It actually means "checking the authorization argument of require() function".

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+          if (expr["kind"] == "functionCall")
+          {
+            // require(.) authorization
+            if (expr["expression"]["nodeType"].get<std::string>() == "Identifier")
```
Try to break this large if-statement into smaller methods.

kunjsong01 replied:
Done. 
Split into the following functions:
* check_authorization_through_tx_origin
* check_require_call
* check_require_argument
* check_tx_origin

lucasccordeiro commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,119 @@
+
+void pattern_checker::print_json_element(const nlohmann::json &json_in)
+{
+  //printf("### %s element[%u] content: key=%s, size=%lu ###\n",
```
Remove.

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/pattern_check.h:
```diff
@@ -0,0 +1,36 @@
+#ifndef SOLIDITY_FRONTEND_PATTERN_CHECK_H_
+#define SOLIDITY_FRONTEND_PATTERN_CHECK_H_
+
+//#define __STDC_LIMIT_MACROS
```
Remove.

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+    std::string node_type = ast_node["nodeType"].get<std::string>();
+    printf("@@ Converting node[%u]: name=%s, nodeType=%s ...\n",
+        index, node_name.c_str(), node_type.c_str());
+    //print_json_array_element(ast_node, node_type, index);
```
Remove.

kunjsong01 replied:
Done.

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+    assert(!"Missing \'nodeType\' filed in ast_node");
+
+  SolidityGrammar::VarDeclStmtT type = SolidityGrammar::get_var_decl_stmt_t(ast_node);
+  printf("	@@@ got Variable-declaration-statement: SolidityGrammar::VarDeclStmtT::%s\n",
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
ACK. I'll replace them with `msg.debug(fmt::format(.))`.

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+// debug functions
+void solidity_convertert::print_json(const nlohmann::json &json_in)
+{
+  printf("### JSON: ###\n");
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
Same reason as explained in the comment of Line 1828 (function `print_json_stmt_element`)

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+void solidity_convertert::print_json_array_element(const nlohmann::json &json_in,
+    const std::string& node_type, const unsigned index)
+{
+  printf("### node[%u]: nodeType=%s ###\n", index, node_type.c_str());
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
Same reason as explained in the comment of Line 1828 (function `print_json_stmt_element`)

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+void solidity_convertert::print_json_element(const nlohmann::json &json_in, const unsigned index,
+    const std::string &key, const std::string& json_name)
+{
+  printf("### %s element[%u] content: key=%s, size=%lu ###\n",
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
Same reason as explained in the comment of Line 1828 (function `print_json_stmt_element`)

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+void solidity_convertert::print_json_stmt_element(const nlohmann::json &json_in,
+    const std::string& node_type, const unsigned index)
+{
+  printf("\t### stmt[%u]: nodeType=%s ###\n", index, node_type.c_str());
```
Why don't we use our messaging interface based on the fmt library?

kunjsong01 replied:
The `print_json_*` functions are debug functions used to print the content of a `nlohmann::json` object (or JSON array element) in LLDB. They are not part of ESBMC's verification flow. The screenshot below shows an example of usage. 

![print_json](https://user-images.githubusercontent.com/32592800/134480845-6cd889b8-cb1e-4994-aed6-d346642ab2b6.png)

lucasccordeiro reviewed (changes requested):
@kunjsong01: could you please add some regression tests to this PR? In particular, you can use the benchmarks from your MSc thesis.

rafaelsamenezes commented on src/esbmc/options.cpp:
```diff
@@ -58,6 +58,9 @@ const struct group_opt_templ all_cmd_options[] = {
     {"warning,W",
      boost::program_options::value<std::vector<std::string>>(),
      ""},
+    {"dummymain,M",
```
I don't think that `dummymain` is a good name. Specially without a good description.

You could also move to a solidity options group

kunjsong01 replied:
Done. 
Fixed in the latest patch.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.h:
```diff
@@ -0,0 +1,36 @@
+//#define __STDC_FORMAT_MACROS
+
+#include <memory>
+#include <iostream>
```
Avoid iostream if possible

kunjsong01 replied:
Done.

rafaelsamenezes reviewed (changes requested):
I will still test out the code, but it looks good overall. I would say:

- More regression tests. They will also help us to evaluate this PR.
- If possible, unit/fuzz tests.
- Remember to add your name/email in the new modules.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert_literals.cpp:
```diff
@@ -51,7 +51,7 @@ bool clang_c_convertert::convert_integer_literal(
     return true;
 
-  assert(type.is_unsignedbv() || type.is_signedbv());
+  assert(type.is_unsignedbv() || type.is_signedbv()); // for "_x=100", false || true. Same as "assert(sum>100)"
```
Why do we need this ` // for "_x=100", false || true. Same as "assert(sum>100)"`?

kunjsong01 replied:
This comment has been removed in previous patches.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert_literals.cpp:
```diff
@@ -63,10 +63,10 @@ bool clang_c_convertert::convert_integer_literal(
       type);
   }
-  else
+  else // "_x=100" uses this. "100" is considered as signed by default. Same as "assert(sum>100)"
```
Why do we need this `// "_x=100" uses this. "100" is considered as signed by default. Same as "assert(sum>100)"`?

kunjsong01 replied:
This comment has been removed in previous patches.

kunjsong01 committed:
Fixes made as per comments:

  - Refactored the large "if" statement in pattern check
  - Removed unused code
  - Replaced status print with msg.status and fmt library
  - Improved comments

kunjsong01 committed:
Fixes made as per comments - Part II:

  - Fixed the name of empty main
  - Added name and email in the new modules

kunjsong01 commented:
I've done the fixes as per comments. 

Two questions regarding regression: 
Q1. Should the test cases be added in `regression/esbmc` or a new directory something like `regression/esbmc-solidity`?
Q2. How do we add multiple test cases for one PR in src/esbmc?

lucasccordeiro commented:
Q1. Should the test cases be added in regression/esbmc or a new directory something like regression/esbmc-solidity?

You can add something like `regression/esbmc-solidity`.

Q2. How do we add multiple test cases for one PR in src/esbmc?

In `regression/esbmc-solidity`, you can create other directories, one for each test case. You can see some examples like `esbmc` or `esbmc-cpp`.

kunjsong01 committed:
Added test cases in regressio/esbmc-solidity

mikhailramalho commented on regression/esbmc-solidity/github_497_6_2/MyContract_satisfiability_2.sol:
```diff
@@ -0,0 +1,59 @@
+    }
+}
+
+
```
Do we need this commented code?

kunjsong01 replied:
Done. 
Removed.

mikhailramalho commented on regression/esbmc-solidity/github_497_6_4/MyContract_satisfiability_4.sol:
```diff
@@ -0,0 +1,60 @@
+    }
+}
+
+
```
Do we need this commented code?

kunjsong01 replied:
Done. 
Removed.

mikhailramalho commented on regression/esbmc-solidity/github_497_7/MyContract_satisfiability_VERIFIER.sol:
```diff
@@ -0,0 +1,63 @@
+}
+
+
+      //__ESBMC_assume(y < 255);
```
Do we need this commented code?

kunjsong01 replied:
Done. Removed.

mikhailramalho commented on src/esbmc/CMakeLists.txt:
```diff
@@ -4,6 +4,12 @@ else()
+  set(SOLIDITY_FRONTEND_TARGETS solidityfrontend)
+else()
+  set(SOLIDITY_FRONTEND_TARGETS "")
+endif()
```
I think by default the variable will be empty if it's not set. Do we need this else branch?

kunjsong01 replied:
Done. 
I removed the `else` block and did a few experiments - it's by default empty. 
Also removed the `else` block in `if(ENABLE_OLD_FRONTEND)`.

mikhailramalho commented on src/langapi/mode.h:
```diff
@@ -42,6 +44,10 @@ languaget *new_cpp_language();
   }
+#define LANGAPI_HAVE_MODE_SOLAST                                               \
+  {                                                                            \
+    "Solidity AST", &new_solidity_language, extensions_sol_ast             \
```
fix code style.

kunjsong01 replied:
Done.
Fixed.

mikhailramalho commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,123 @@
+  const nlohmann::json &body_stmt = func["body"]["statements"];
+  msg.progress("  - Pattern-based checking: SWC-115 Authorization through tx.origin");
+  printf("statements in function body array ... \n");
+  //print_json_element(body_stmt);
```
remove printf.

mikhailramalho commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,123 @@
+  nlohmann::json::const_iterator itr = body_stmt.begin();
+  for (; itr != body_stmt.end(); ++itr, ++index)
+  {
+    msg.status(fmt::format("@@ checking function body stmt {}", index));
```
why use the @ here?

kunjsong01 replied:
Done. 
Removed.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+#include <util/mp_arith.h>
+#include <util/std_code.h>
+#include <util/std_expr.h>
+#include <iomanip>
```
do you need iomanip here?

kunjsong01 replied:
Done. 
Removed. The code is now using msg and fmt interfaces.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+  if (is_array)
+  {
+    assert(!"Unimplemented - funciton parameter is array type");
+    //param_type.id("pointer");
```
Remove comment.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+      locationt location_end;
+      get_start_location_from_stmt(block, location_end);
+
+      //assert(!"done - all CompoundStmtClass?");
```
Remove comment.

kunjsong01 replied:
Done.
Removed.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+{
+  // For rule expression
+  // We need to do location settings to match clang C's number of times to set the locations when recurring
+  static int call_expr_times = 0; // TODO: remove debug
```
Remove debug code.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+
+  std::string name, id;
+  name = decl["name"].get<std::string>();
+  id = "c:@F@" + name;
```
Do we need this prefix?

kunjsong01 replied:
Done. 
Removed.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+  {
+    if(context.move(symbol, s))
+    {
+      std::cerr << "Couldn't add symbol " << symbol.name
```
you need to use the message interface here.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1831 @@
+    const std::string& node_type, const unsigned index)
+{
+  printf("\t### stmt[%u]: nodeType=%s ###\n", index, node_type.c_str());
+  std::cout << std::setw(2) << json_in << '\n'; // '2' means 2x indentations in front of each line
```
Don´t use cout.

mikhailramalho commented on src/solidity-frontend/solidity_convert.h:
```diff
@@ -0,0 +1,113 @@
+#define SOLIDITY_FRONTEND_SOLIDITY_CONVERT_H_
+
+#include <memory>
+#include <iostream>
```
Don´t use iostream, you should use the message interface.

kunjsong01 replied:
Done. 
Removed. All printf, cout and cerr calls have been replaced by the unified msg interface.

kunjsong01 replied:
Done. 
Removed. The code is now using the msg and fmt interfaces.

mikhailramalho commented on src/solidity-frontend/solidity_grammar.h:
```diff
@@ -0,0 +1,190 @@
+#include <string>
+#include <nlohmann/json.hpp>
+
+#define ENUM_TO_STR(s) case s: { return #s; }
```
Does this macro have to be defined on this header? Can you move it to the cpp file where it's used?

kunjsong01 replied:
Done.
Moved to the corresponding cpp file.

mikhailramalho commented on src/solidity-frontend/solidity_language.cpp:
```diff
@@ -0,0 +1,133 @@
+
+solidity_languaget::~solidity_languaget()
+{
+  if(clang_c_module != nullptr)
```
you probably don't need this here.

kunjsong01 replied:
Done. 
Changed to `assert` to make sure we have valid `clang_c_module`.

mikhailramalho commented on src/solidity-frontend/solidity_language.cpp:
```diff
@@ -0,0 +1,133 @@
+  while (getline(ast_json_file_stream, new_line))
+  {
+    // file pointer continues from "=== *.sol ==="
+    //printf("new_line: %s\n", new_line.c_str());
```
remove commented code from this file.

kunjsong01 replied:
Done. 
Removed.

mikhailramalho reviewed (commented):
Hi @kunjsong01, thanks for the PR. Some good things I've found:
* Great amount of comments.
* Good initial number of tests.

I've left a few comments about stuff to improve here and there before we can accept the PR, in particular:
1. The new `-M` flag seems a bit weird... Do we actually need it? The cpp programs all seem empty with only a main returning 0.
2. Why do we need the convert_intrinsics in the clang frontend?
3. The code to select the new frontend, the mode, etc. also seem weird, but I'll clone and try your PR before I can give more comments on that.
4. You need to remove all calls to cerr, cout and printf, and to use our unified message interface.
5. You need to remove all debug code, including variables you used for testing or code that is commented.

Some minor comments that won't prevent us from accepting your PR but it would be great if you could fix it:
1. Remove the pattern where the `for` loop initialization is empty, e.g., [here](https://github.com/esbmc/esbmc/pull/497/files#diff-31c440873e3d1918403518ce8a66621fb8399b230c7fe40d3a3a4b1fcd5c7840R89).
2. No need to add curly braces when `if` or `case` is a one-liner.

Overall, I think it's a good contribution and it needs a small amount of work to get everything ready to merge.

kunjsong01 committed:
Fixes made as per comments - Part III:

  - Replaced cerr/cout with unified msg interface
  - Removed unused debug code and variables
  - Removed curly braces for one-line if or case
  - Fixed for loops that have empty initializations
  - Replaced printf with unified msg interface
  - Also fixed regex in esbmc-solidity regression

kunjsong01 committed:
Fixes made as per comments - Part IV:

  - More cosmetic fixes
  - Minor CMake refactor
  - Removed unnecessary prefixes in function symbol ID

kunjsong01 commented:
The following patches addressed all issues in the comments:
* Fixes made as per comments - Part III 
* Fixes made as per comments - Part IV

100% pass in ebm-solidity regression. 

Regression log: 
[ebm_solidity_regression.log](https://github.com/esbmc/esbmc/files/7244236/ebm_solidity_regression.log)

----------------------------------------------------------------
Q&A:

> 1.The new -M flag seems a bit weird... Do we actually need it? The cpp programs all seem empty with only a main returning 0.

Yes, I think we need `-M` to make clang_c_module work. Otherwise, it would fail in line 213 when making a call to `buildASTs` in function:  
`clang_c_languaget::parse`
```cpp
212   // Generate ASTUnit and add to our vector
213   auto AST = buildASTs(intrinsics, new_compiler_args);
```
with the following error message: 
```
error: no input files
Assertion failed: (Jobs.size() == 1), function buildASTs, file ../src/clang-c-frontend/AST/build_ast.cpp, line 90.
```

'-M' is a quick workaround to resolve this error. It may not be the most elegant solution. Perhaps, would it be possible to have this PR merged and mark it as future improvement? 

> 2. Why do we need the convert_intrinsics in the clang frontend?

There are three steps in `clang_c_languaget::typecheck`:
1. convert()
2. adjust()
3. c_link()

`convert_intrinsics` just does Step 1. It's used by the `clang_c_module` in Solidity frontend to add symbols of the intrinsic declarations to the context without adjusting or c_link the context. We need these intrinsics to support basic operations - assert, assume and nondet, as well as SV-COMP declarations (e.g. \_\_VERIFIER\_\_).

mikhailramalho commented:
Em ter., 28 de set. de 2021 às 10:39, KSong ***@***.***>
escreveu:

> The following patches addressed all issues in the comments:
>
>    - Fixes made as per comments - Part III
>    - Fixes made as per comments - Part IV
>
> 100% pass in ebm-solidity regression.
>
> Regression log:
> ebm_solidity_regression.log
> <https://github.com/esbmc/esbmc/files/7244236/ebm_solidity_regression.log>
> ------------------------------
>
> Q&A:
>
> 1.The new -M flag seems a bit weird... Do we actually need it? The cpp
> programs all seem empty with only a main returning 0.
>
> Yes, I think we need -M to make clang_c_module work. Otherwise, it would
> fail in line 213 when making a call to buildASTs in function:
> clang_c_languaget::parse
>
> 212   // Generate ASTUnit and add to our vector213   auto AST = buildASTs(intrinsics, new_compiler_args);
>
> with the following error message:
>
> error: no input files
> Assertion failed: (Jobs.size() == 1), function buildASTs, file ../src/clang-c-frontend/AST/build_ast.cpp, line 90.
>
> '-M' is a quick workaround to resolve this error. It may not be the most
> elegant solution. Perhaps, would it be possible to have this PR merged and
> mark it as future improvement?
>

I prefer we don't merge unfinished work, historically these FIXMEs are
never removed/finished in ESBMC and are left to bit rot.

My first question is if you try to remove this assertion?


>    1. Why do we need the convert_intrinsics in the clang frontend?
>
> There are three steps in clang_c_languaget::typecheck:
>
>    1. convert()
>    2. adjust()
>    3. c_link()
>
> convert_intrinsics just does Step #1
> <https://github.com/esbmc/esbmc/pull/1>. It's used by the clang_c_module
> in Solidity frontend to add symbols of the intrinsic declarations to the
> context without adjusting or c_link the context. We need these intrinsics
> to support basic operations - assert, assume and nondet, as well as SV-COMP
> declarations (e.g. *VERIFIER**).
>

I see, this explains the non-modular design of the solution.

How about you inherit from clang_c_frontend and override the behaviour that
doesn't make sense in your new frontend? Similar to how it's done for the
clang c++ frontend. For instance, the c_link is specially important if you
are trying to verify more than one file.

Also, are you using the headers in regression/esbmc-solidity/, e.g,
apache.h? If you aren't, please remove them.


> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/497#issuecomment-929294746>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJHY462D7K4CY25BTMTTUEHHSJANCNFSM5EM3OSDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 

Mikhail Ramalho.

kunjsong01 commented:
> How about you inherit from clang_c_frontend and override the behaviour that
doesn't make sense in your new frontend?

Thanks very much for the comments. 
Agreed. I'll make Solidity frontend inherits from clang_c_frontend to remove "-M" and "convert_intrincs".

kunjsong01 committed:
Fixes made as per comments - Part V:

  - Make new frontend inherit from clang-c-frontend
  - Removed convert_intrinsics
  - Removed header files in Solidity frontend regression, e.g. apache.h

kunjsong01 commented:
The last patch contains the improvements as following:
  - Make new frontend inherit from clang-c-frontend
  - Removed convert_intrinsics
  - Removed header files in Solidity frontend regression, e.g. apache.h

As for the "-M main.c" argument, it's a bit tricky to remove. Because It serves as the key component to make `clang_c_languaget::parse` work, which is a prerequsite to get the symbols of intrinsics. 

> My first question is if you try to remove this assertion?

I removed the assertion but it failed with "Segmentation fault" in line 93 of `src/clang-c-frontend/AST/build_ast.cpp`: 
```cpp
 89   const clang::driver::JobList &Jobs = Compilation->getJobs();
 90   printf("@@ This is Jobs.size(): %d\n", Jobs.size());
 91   //assert(Jobs.size() == 1); 
 92
 93   const llvm::opt::ArgStringList *const CC1Args = &Jobs.begin()->getArguments();
```

Because Jobs.size() == 0. This appears to be blocking issue to this PR. 

Question: 
Q1. The `buildASTs` function of `src/clang-c-frontend/AST/build_ast.cpp` --- is there a way to use this function without any compilation jobs? 

@mikhailramalho Could you please shed some light?

rafaelsamenezes commented on src/esbmc/options.cpp:
```diff
@@ -46,6 +46,12 @@ const struct group_opt_templ all_cmd_options[] = {
     {"show-symex-value-sets",
      NULL,
      "show value-set analysis during symbolic execution"}}},
+  {"Solidity frontend",
```
Put an ifdef for solidity in this block

kunjsong01 replied:
Done.
Removed along with "-M".

rafaelsamenezes commented on src/langapi/language_ui.cpp:
```diff
@@ -13,6 +13,9 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 #include <util/i2string.h>
 #include <util/show_symbol_table.h>
 
+extern std::string sol_main;
```
Can't this me initialized inside the object?

kunjsong01 replied:
Done.
Removed along with "-M".

rafaelsamenezes commented on regression/esbmc-solidity/Makefile:
```diff
@@ -0,0 +1,19 @@
+default: tests.log
```
You don't need this file. Add the directory here: https://github.com/esbmc/esbmc/blob/master/regression/CMakeLists.txt

kunjsong01 replied:
Done.

kunjsong01 committed:
Fixes made as per comments - Part VI:

  - Removed -M flag
  - Updated regression script in esbmc-solidity

kunjsong01 commented:
"-M" flag is removed. Solidity frontend is now using a temporary file as the input to clang-tool. I implemented the code in this way becuase it appears impossible to use `clang_c_languaget::parse` and `buildASTs` functions without giving a valid path to a source C file. 

![image](https://user-images.githubusercontent.com/32592800/136196504-4993ef05-316a-40af-ace7-4f244fe55107.png)

Now we can just use the cmd line without "-M" flag: 
`esbmc $CONTRACT_DIR/swc-101/MyContract_overflow.solast --function func_overflow --compact-trace`

Got 100% pass in regression: 
![image](https://user-images.githubusercontent.com/32592800/136196116-6f03682f-10dd-43e5-982e-5df04f2cab6b.png)

-----------------

On a side note, the whole point of using clang-tool in Solidity frontend is to add the symbols of intrinsics to the context (i.e. the symbol table). The intrinsic declarations are handled by `clang-c-forntend`, and the Solidity declarations are handled by the `solidity-frontend`. 
![image](https://user-images.githubusercontent.com/32592800/136197130-66c8bcff-a2f0-4dc9-b86f-12dcca60d6e4.png)

kunjsong01 committed:
Fixed github workflows

kunjsong01 committed:
Fixed build errors for -DENABLE_WERROR=On

kunjsong01 committed:
Fixes to pass Github workflows:

  - Fixed build errors in Windows regression
  - Replaced JSON lib with the lightweight version
  - Fixed the temp file problem in Linux regression
  - Fixed Lint errors

lucasccordeiro reviewed (approved):
LGTM.

rafaelsamenezes commented on regression/esbmc-solidity/Makefile:
```diff
@@ -0,0 +1,17 @@
+clean:
+	find -name '*.out' -execdir $(RM) '{}' \;
+	find -name '*.gb' -execdir $(RM) '{}' \;
+	$(RM) tests.log
```
You can remove this file

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/langapi/language_ui.cpp:
```diff
@@ -13,6 +13,8 @@ Author: Daniel Kroening, kroening@cs.cmu.edu
 #include <util/i2string.h>
 #include <util/show_symbol_table.h>
 
+extern std::string sol_function;
```
I don't like this as an extern/global. From what I am looking, this variable is initialized during the cmdline processing and
afterwards is used on the language_uit. I suggest:

1. Removing this global variable
2. Before the `language.set_sol_func` you use the `cmdlinet` object that you have in the scope to fetch the argument.

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,124 @@
+
+bool pattern_checker::pattern_checker::do_pattern_check()
+{
+  // TODO: add more functions here to perform more pattern-based checks
```
Remember to open issues for this TODO

kunjsong01 replied:
ACK. 
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.h:
```diff
@@ -0,0 +1,37 @@
+#include <nlohmann/json.hpp>
+#include <solidity-frontend/solidity_grammar.h>
+
+class pattern_checker
```
Add a comment explaining the purpose of this class

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.h:
```diff
@@ -0,0 +1,37 @@
+    const messaget &msg);
+  virtual ~pattern_checker() = default;
+
+  bool do_pattern_check();
```
What the return value means in this context?

kunjsong01 replied:
Similar to the boolean values returned in other frontend, e.g.: 
```cpp
bool clang_c_convertert::convert()
```
'false' means everything is fine. 'true' means something went wrong.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.h:
```diff
@@ -0,0 +1,37 @@
+  virtual ~pattern_checker() = default;
+
+  bool do_pattern_check();
+  bool start_pattern_based_check(const nlohmann::json &func);
```
What the return value means in this context?

kunjsong01 replied:
Similar to the boolean values returned in other frontend, e.g.: 
```cpp
bool clang_c_convertert::convert()
```
'false' means everything is fine. 'true' means something went wrong.

rafaelsamenezes commented on src/solidity-frontend/pattern_check.cpp:
```diff
@@ -0,0 +1,124 @@
+  // look for BinaryOperation "=="
+  if(arg_expr["nodeType"].get<std::string>() == "BinaryOperation")
+  {
+    if(arg_expr["operator"].get<std::string>() == "==")
```
Is this the only binary operator that is used? If not, then add a TODO.

kunjsong01 replied:
Yes, function 'check_require_call' is part of the pattern check for `Authorization via TxOrigin`. 
`==` is the only binary operator used in this check.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  assert(found_contract_def);
+
+  // reasoning-based verification
+  //assert(!"Continue with symbol annotations");
```
Remove this lijne

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      return true;
+  }
+
+  assert(current_functionDecl == nullptr);
```
Could you explain this assertion?

kunjsong01 replied:
Done. 
A comment has been added in the code.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  // However, ExpressionStatement node just contains "typeDescriptions".
+  // For consistensy, we use ["typeName"]["typeDescriptions"] as in state-variable-declaration
+  // to improve the re-usability of get_type* function, when dealing with non-array var decls.
+  // For array, do NOT use ["typeName"]. Otherwise, it will cause problem
```
I didn't follow. Do you mean in this function or during the parsing phase?

If it is during this function, is there a way to convert this comment into an assertion?

kunjsong01 replied:
> Do you mean in this function or during the parsing phase?

It isn't about the parsing phase. Everything in `solidity_convert.cpp` is associated with typechecking phase.

> If it is during this function, is there a way to convert this comment into an assertion?

I don't think it can be converted into an assertion. Because the whole purpose of these comments is to remind people why using ["typeName"]["typeDescription"] instead of other sources type information in Solidity JSON AST. By "other sources", I mean: 
![image](https://user-images.githubusercontent.com/32592800/137314736-4221fe52-fab6-4a33-b896-f11a66ac81b5.png)

I believe comments are good enough here.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+
+  symbol.lvalue = true;
+  symbol.is_extern =
+    false; // TODO: hard coded for now, may need to change later
```
Remember to open an issue for this.

kunjsong01 replied:
ACK. 
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  get_start_location_from_stmt(location);
+
+  SolidityGrammar::BlockT type = SolidityGrammar::get_block_t(block);
+  msg.status(fmt::format(
```
maybe msg.debug instead?

kunjsong01 replied:
Done. 
Replaced msg.status with msg.debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      _block.operands().push_back(statement);
+      ++ctr;
+    }
+    msg.status(fmt::format(" \t@@@ CompoundStmt has {} statements", ctr));
```
maybe msg.debug instead?

kunjsong01 replied:
Done. 
Replaced msg.status with msg.debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    unsigned ctr = 0;
+    // N.B. Although Solidity AST JSON uses "declarations": [],
+    // there will ALWAYS be 1 declaration!
+    // A second declaration will become another stmt in "statements"
```
Can you add an example in the comment. I didn't follow what you meant

kunjsong01 replied:
Done. 
Updated the comment to include an example.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      decls.operands().push_back(single_decl);
+      ++ctr;
+    }
+    msg.status(fmt::format(" \t@@@ DeclStmt group has {} decls", ctr));
```
maybe msg.debug instead?

kunjsong01 replied:
Done. 
Replaced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      assert(!"Error: ESBMC could not find the parent scope for this ReturnStatement");
+
+    // 1. get return type
+    // TODO: Fix me! Assumptions:
```
Is this coment still valid?

kunjsong01 replied:
Yes. When a new benchmark fails the assertions below, these comments provide more details for debugging and understanding the root cause. 

In general, the `solidity_convert.cpp` is coded in a defensive way - it uses some assertions to make sure: 
1. The input JSON AST is well-formed.
2. Each function is used to deal with one type of AST node. 
3. If #1 and #2 are satisfied, it will abort in the earlist phase.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  case SolidityGrammar::StatementT::ForStatement:
+  {
+    // Based on rule for-statement
+    // TODO: Fix me. Assuming for loop contains everything
```
Remember to open an issue for this

kunjsong01 replied:
ACK. 
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    assert(
+      stmt.contains("initializationExpression") && stmt.contains("condition") &&
+      stmt.contains("loopExpression") && stmt.contains("body"));
+    // TODO: Fix me. Currently we don't support nested for loop
```
Remember to open an issue for this

kunjsong01 replied:
ACK. 
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  get_start_location_from_stmt(location);
+
+  SolidityGrammar::ExpressionT type = SolidityGrammar::get_expression_t(expr);
+  msg.status(fmt::format(
```
msg.debug?

kunjsong01 replied:
Done. 
Repalced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      call.arguments().push_back(single_arg);
+      ++num_args;
+    }
+    msg.status(fmt::format("  @@ num_args={}", num_args));
```
msg.debug?

kunjsong01 replied:
Done. 
Replaced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      return true;
+  }
+  else
+  {
```
Remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  // For "Assignment" expression, it's called "leftHandSide" or "rightHandSide".
+  // For "BinaryOperation" expression, it's called "leftExpression" or "leftExpression"
+  exprt lhs, rhs;
+  if(expr.contains("leftHandSide"))
+  {
+    if(get_expr(expr["leftHandSide"], lhs))
+      return true;
+
+    if(get_expr(expr["rightHandSide"], rhs))
+      return true;
+  }
+  else if(expr.contains("leftExpression"))
+  {
+    if(get_expr(expr["leftExpression"], lhs))
+      return true;
+
+    if(get_expr(expr["rightExpression"], rhs))
+      return true;
+  }
```
I think that you can simplify this with ternary operators and moving the assertion before e.g

```cpp
assert(expr.contains("leftHandSide") || expr.contains("leftExpression") );
if(get_expr(expr.contains("leftHandSide") ? expr["leftHandSide"] : expr["leftExpression"], lhs))
      return true;
```

kunjsong01 replied:
Thanks for this comment. Using the ternary operator certainly can make the code more compact. 
```cpp
  exprt lhs, rhs;
  assert(expr.contains("leftHandSide") || expr.contains("leftExpression"));
  if(get_expr(expr.contains("leftHandSide") ? expr["leftHandSide"] : expr["leftExpression"], lhs))
      return true;
  if(get_expr(expr.contains("leftHandSide") ? expr["rightHandSide"] : expr["rightExpression"], rhs))
      return true;
```

However, it would have lost the readibility because the "left*" and "right*" seem pretty confusing in this way. 
It would have also lost the usage of assert(!"<message>"), which was palced in the 'else' block to accelerate the debugging process when extending the new frontend to support more benchamrks.

The new frontend is an early prototype, and it is still subject to further extensions. I would not trade readability for compactness for the time being.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  // 3. Convert opcode
+  SolidityGrammar::ExpressionT opcode =
+    SolidityGrammar::get_expr_operator_t(expr);
+  msg.status(fmt::format(
```
msg.debug?

kunjsong01 replied:
Done. 
Replaced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  typet uniop_type;
+  assert(
+    expr
+      ["prefix"]); // TODO: Fix me! Currently just support prefix == true,e.g. pre-increment
```
Remember to open an issue

kunjsong01 replied:
ACK. 
will do it when extending new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  // 3. get UnaryOperation opcode
+  SolidityGrammar::ExpressionT opcode =
+    SolidityGrammar::get_expr_operator_t(expr, expr["prefix"]);
+  msg.status(fmt::format(
```
msg.debug

kunjsong01 replied:
Done. 
Replaced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+
+    if(
+      sub_type.is_struct() ||
+      sub_type.is_union()) // for "assert(sum > 100)", false || false
```
remove the comment

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    if(
+      get_array_to_pointer_type(
+        type_name,
+        sub_type)) // TODO: Fix me! Inconsistency. Better to use Solidity compiler library?
```
Remember to open an issue

kunjsong01 replied:
ACK. 
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+
+    if(
+      sub_type.is_struct() ||
+      sub_type.is_union()) // for "assert(sum > 100)", false || false
```
remove the comment

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  }
+  default:
+  {
+    msg.status(fmt::format(
```
msg.debug?

kunjsong01 replied:
Done. 
Replaced status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  // TODO: More var decl attributes checks:
+  //    - Constant
+  //    - Volatile
+  //    - isRestrict
```
Remember to open an issue

kunjsong01 replied:
ACK.
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  //    - Constant
+  //    - Volatile
+  //    - isRestrict
+  return false;
```
remember to open an issue

kunjsong01 replied:
ACK.
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    new_type.set("#cpp_type", "unsigned_char");
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  //    - Constant
+  //    - Volatile
+  //    - isRestrict
+  return false;
```
remember to open an issue

kunjsong01 replied:
ACK.
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  }
+  default:
+  {
+    msg.status(fmt::format(
```
msg.debug?

kunjsong01 replied:
Done. 
Replace status with debug.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  {
+    assert(
+      type_name["parameters"].size() ==
+      1); // TODO: Fix me! assuming one return parameter
```
Remember to open an issue.

kunjsong01 replied:
ACK.
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      assert(!"Unable to find the corresponding local variable decl. Function body  does not have statements.");
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      }
+    }
+    else
+    {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    }
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  if(
+    ast_node["name"] == "__ESBMC_assume" ||
+    ast_node["name"] == "__VERIFIER_assume")
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+  {
+    return true;
+  }
+  else
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    return false;
+  }
+
+  return false; // make old compiler happy, e.g. pre-GCC4.9
```
Can't you just return the if statement? e.g:

```c
return ast_node["name"] == "__ESBMC_assume" ||  ast_node["name"] == "__VERIFIER_assume";
```

kunjsong01 replied:
Done. 
Replaced the if-else block with one line return.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      {
+        // e.g. for typeString like:
+        // "typeString": "function () returns (uint8)"
+        // TODO: currently we only assume one parameters
```
Remember to open an issue

kunjsong01 replied:
ACK.
will do it when extending the new frontend to support more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+          adjusted_expr["returnParameters"] = j2;
+        }
+        else
+        {
```
Remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+      }
+    }
+    else
+    {
```
Remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    }
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+        adjusted_expr = j2;
+      }
+      else
+      {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    }
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    elementary_type = j;
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -0,0 +1,1898 @@
+    the_size = sub_match.str();
+  }
+  else
+  {
```
remove {

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_convert_literals.cpp:
```diff
@@ -0,0 +1,45 @@
+// TODO: String literal
+// TODO: Float literal.
+//    - Note: Currently Solidity does NOT support floating point data types or fp arithmetic.
+//      Everything is done in fixed-point arithmetic as of Solidity compiler v0.8.6.
```
Remember to open an issue

kunjsong01 replied:
ACK.
will do it when supporting more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_grammar.h:
```diff
@@ -0,0 +1,189 @@
+  // TODO: fixed-bytes
+  // TODO: fixed
+  // TODO: ufixed
+  ElementaryTypeNameTError
```
Remember to open an issue

kunjsong01 replied:
ACK.
will do it when supporting more benchmarks.

rafaelsamenezes commented on src/solidity-frontend/solidity_language.cpp:
```diff
@@ -0,0 +1,158 @@
+  add_cprover_library(context, msg);
+  return clang_main(context, msg);
+  //assert(!"come back and continue - solidity_languaget::final");
+  return false;
```
This code is unreachable. Remove it.

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/solidity-frontend/solidity_language.cpp:
```diff
@@ -0,0 +1,158 @@
+  return false;
+}
+
+std::string solidity_languaget::temp_c_file()
```
I know that you are using so that Clang has a compilation job. But I haven't seem where you explained this.

kunjsong01 replied:
Done. 
Added a comment in this function to explain its purpose.

rafaelsamenezes commented on src/solidity-frontend/solidity_language.h:
```diff
@@ -0,0 +1,67 @@
+
+  languaget *clang_c_module;
+
+  //protected:
```
remove this comment

kunjsong01 replied:
Done.

rafaelsamenezes commented on src/util/language.h:
```diff
@@ -82,6 +82,26 @@ class languaget
+  inline std::string get_sol_func()
+  {
+    return sol_func_path;
+  };
```
This is not the place to add this. Maybe in `solidity_languaget` class?

kunjsong01 replied:
Done. 
Solution: Replaced by a general member `func_name` for the verification that requires the function name before GOTO conversion phase, e.g. Solidity.

rafaelsamenezes reviewed (changes requested):
It is almost ready for merge. Most of my requests are for better comments and code-style fixes.

kunjsong01 committed:
Fixes made as per comments - Part VII:

  - Fixed code style and comments
  - Removed extern variable "sol_function"
  - Removed Solidity stuff in class languaget

rafaelsamenezes committed:
Delete Makefile

rafaelsamenezes committed:
removed uneeded brackets

rafaelsamenezes reviewed (approved):
LGTM

mikhailramalho commented on src/solidity-frontend/typecast.h:
```diff
@@ -0,0 +1,10 @@
+#include <util/namespace.h>
+#include <util/std_expr.h>
+
+extern void
```
doesn't have to be extern.

mikhailramalho reviewed (commented):
@kunjsong01 it looks much better, but could you just remove that unnecessary `extern`? 

BTW, since you just implemented the frontend, maybe you could write some documentation about the process and design decisions and add it to github's wiki? It might be helpful in the future.
